### ISSUE TITLE
Overlap Removal Inputs and Outputs

### ISSUE THREAD

**Issue by JamesSaxon** [UCATLAS/xAODAnaHelpers#287]
Hi @kratsg, @mmilesi, all

Per our chat on the overlap removal, I think this is the behavior I would expect, _if one had all input algorithms defined._  I am assuming that the vector\<string\>'s contain "" entries for nominal, and can figure all this out.

Simplifying to electrons and jets for now -- 

  - No systematics are run.  The overlap removal works normally on the nominal containers.
  - A single electron OR jet systematic is defined, while the other is nominal.  The tool runs the single systematic.
  - A single electron systematic AND a single jet systematic are defined.  The nominal jets are missing for the electron variations and vice versa.  The tool should complain loudly if not halt.
  - A single electron systematic is defined, while the jet systematic is "All" (or vice versa, etc.).  It's all fine and dandy for the electrons variation -- nominal exists for the jets.  But the tool should stop or complain because it cannot do the jet variations with nominal electrons.
  - Both electrons and jets have "All" systematics defined.  All jet and electron variations and nominal are provided.

This said, now that I understand better the intended use, I'm fine just commenting out the single line of my configuration.  Let me know what you think best.  :-)

Jamie

**Comment by mmilesi** [comment]
Hi @JamesSaxon , @kratsg 

replies inline:

> Per our chat on the overlap removal, I think this is the behavior I would expect, if one had all input algorithms defined. I am assuming that the vector<string>'s contain "" entries for nominal, and can figure all this out.
> 
> Simplifying to electrons and jets for now --
> 
> No systematics are run. The overlap removal works normally on the nominal containers.

Correct.

> A single electron OR jet systematic is defined, while the other is nominal. The tool runs the single systematic.

Correct. In general, this is the case where user runs systematics on **one object type only** (can be a single syst, e.g., `EG_SCALE_LARCALIB` - and in this case the nominal electrons will not be created at all - or `All` systematics - and in this case the nominal will be included, too), and only the nominal for all the other objects. This is in general what the workflow of the analysis should be, if I am not mistaken.

> A single electron systematic AND a single jet systematic are defined. The nominal jets are missing for the electron variations and vice versa. The tool should complain loudly if not halt.

As you noticed, this will generate a crash, preceded by a lot of alarming messages. But this, based on what I said before, is right what we want (right?).

> A single electron systematic is defined, while the jet systematic is "All" (or vice versa, etc.). It's all fine and dandy for the electrons variation -- nominal exists for the jets. But the tool should stop or complain because it cannot do the jet variations with nominal electrons.

This case is pretty much as the one above: this is not how things should be done.

> Both electrons and jets have "All" systematics defined. All jet and electron variations and nominal are provided.

I haven't tried this case yet, but I assume it will not give problems, although again, it's not how the workflow is intended to be.

> This said, now that I understand better the intended use, I'm fine just commenting out the single line of my configuration. Let me know what you think best. :-)

I will show you what one would get from xAH when trying to run a single syst (EG_SCALE_LARCALIB) **only** for the electrons.

In a typical situation where the order of call would be:

```
JetCalibrator
ElectronCalibrator
JetSelector
ElectronSelector
OverlapRemover
...
```

This is the relevant configuration for:

1. ElectronCalibrator

```
m_inputAlgoSystNames = ""
m_outputAlgoSystNames = "ElectronCalibrator_Syst"
```

2. ElectronSelector

```
m_inputAlgoSystNames = "ElectronCalibrator_Syst"
m_outputAlgoSystNames = "ElectronSelector_Syst"
```
3. OverlapRemover

```
m_inputAlgoElectrons = "ElectronSelector_Syst"
m_outputAlgoElectrons = "ElectronsOR_Syst"

m_inputAlgoJets = ""
m_outputAlgoJets = "AntiKt4EMTopoJetsOR_Syst"

m_inputAlgoMuons = ""
m_outputAlgoMuons = "MuonOR_Syst"

```

(notice that the input algos for jets and muons are empty!!)

By doing like this, the following is the set of object containers the user will end up having in the TStore, and that will be available to the xAH/user algorithm downstream:

```
xAOD::TStore::print       INFO    Contents of transient store:
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_Calib
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Jet_v1> >
xAOD::TStore::print       INFO        Pointer: 0x14561650
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_CalibShallowCopy
xAOD::TStore::print       INFO        Type: DataVector<xAOD::Jet_v1>
xAOD::TStore::print       INFO        Pointer: 0x145893d0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_CalibShallowCopyAux.
xAOD::TStore::print       INFO        Type: xAOD::ShallowAuxContainer
xAOD::TStore::print       INFO        Pointer: 0x1448d6c0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_OREG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Jet_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16a40ae0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_OREG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Jet_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16f406d0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: AntiKt4EMTopoJets_Selected
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Jet_v1> >
xAOD::TStore::print       INFO        Pointer: 0x13e6ed70
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: ElectronCalibrator_Syst
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x163458d0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: ElectronSelector_Syst
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x17513da0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: ElectronsOR_Syst				
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x14997660
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibEG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x1492a020
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibEG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16a45a70
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibShallowCopyAux.EG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: xAOD::ShallowAuxContainer
xAOD::TStore::print       INFO        Pointer: 0x17576ce0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibShallowCopyAux.EG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: xAOD::ShallowAuxContainer
xAOD::TStore::print       INFO        Pointer: 0x13964820
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibShallowCopyEG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: DataVector<xAOD::Electron_v1>
xAOD::TStore::print       INFO        Pointer: 0x17578c00
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_CalibShallowCopyEG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: DataVector<xAOD::Electron_v1>
xAOD::TStore::print       INFO        Pointer: 0x16f40440
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_OREG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16b083a0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_OREG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x13e75900
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_SelectedEG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x13961b30
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Electrons_SelectedEG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Electron_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16a40d10
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: JetCalibrator_Syst
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x1383f0b0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: MuonCalibrator_Syst
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x15fd47d0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: MuonSelector_Syst
xAOD::TStore::print       INFO        Type: vector<string>
xAOD::TStore::print       INFO        Pointer: 0x164076e0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_Calib
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Muon_v1> >
xAOD::TStore::print       INFO        Pointer: 0x17069c40
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_CalibShallowCopy
xAOD::TStore::print       INFO        Type: DataVector<xAOD::Muon_v1>
xAOD::TStore::print       INFO        Pointer: 0x159258c0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_CalibShallowCopyAux.
xAOD::TStore::print       INFO        Type: xAOD::ShallowAuxContainer
xAOD::TStore::print       INFO        Pointer: 0x16b067f0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: Yes
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_OREG_SCALE_LARCALIB__1down
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Muon_v1> >
xAOD::TStore::print       INFO        Pointer: 0x16a409e0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_OREG_SCALE_LARCALIB__1up
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Muon_v1> >
xAOD::TStore::print       INFO        Pointer: 0x13e75a00
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
xAOD::TStore::print       INFO      Name: Muons_Selected
xAOD::TStore::print       INFO        Type: ConstDataVector<DataVector<xAOD::Muon_v1> >
xAOD::TStore::print       INFO        Pointer: 0x173801d0
xAOD::TStore::print       INFO        IsOwner: Yes
xAOD::TStore::print       INFO        HasDictionary: No
xAOD::TStore::print       INFO    -----------------------------------
```

As you can see, the containers after the Overlap Removal (`*_OR*`) will have a name identifier with the systematic name appended to it. So, any algorithm downstream should be made able to pick them up the correct way,  by reading the systematics names from, e.g.,  `vector<string> ElectronOR_Syst`, which are stored in TStore as well. Of course, one could simply configure the downstream algos, e.g. the tree manager, using the **exact** name of the desired container.  

To my knowledge, at this stage of xAH implementation, nobody have really tried to think about how to dump all this systematically varied containers to a persistent output , i.e., a `TTree` or a mini-xAOD. For the tree case, it won't be too difficult to modify `TreeAlgo`, or any tree manager class derived from it that has access to the TStore, to let it read the different containers and dump separate trees for each syst. Maybe @kratsg can chime in on this.

I would really appreciate any comments, ideas, suggestions to make this the most flexible and easy for the user to handle!

Cheers

Marco

**Comment by JamesSaxon** [comment]
Hi Marco -- 

I just tried what you suggested, with a [simple a configuration](https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/VBFInv/AnalysisCode/trunk/scripts/syst_test.py) based on [these dictionaries](https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/VBFInv/AnalysisCode/trunk/scripts/vbfinv_baseline.py).

I finally understand how this works, but I find the current setup quite confusing, and it requires a lot of changes in order to turn on or of a systematic.  If I want to turn on an electron systematic, I have to 

   1. Turn on the systematic.
   2. Connect the electron input algo to the OR.
   3. Connect the downstream algorithms to ElectronORAlgo

I think it would make a _lot_ more since that all of the downstream algorithms remain connected -- jets to the ORAlgoJet, muons to ORAlgoMuon, electrons to ORAlgoElectrons.  This, in fact, is how you wrote the example, but as you see in your example, only the ```ElectronsOR_Syst``` vector\<string\> shows up.

Actually, personally, I think it would make most sense if one could simply turn off or on the systematics, _always_ leaving the algorithms "threaded" together.  If more than one of the input vectors did not contain "" (nominal), the OR would complain or fail.  I don't think that's so difficult to check for.

What do you think?  

Thanks -- 

Jamie

**Comment by kratsg** [comment]
Hi @JamesSaxon,

Unfortunately, this removes from the flexibility of the code. Imagine for example, two different kinds of electrons you want to look at -- rather than creating two sets of jobs -- you manage your two sets of electrons and weave them in and out of the various systematic calibrations.

If you only want to change the set of systematics, you only need to change that first `m_systName` that @mmilesi pointed out -- without changing any of the other names and leaving it all threaded together.

**Comment by JamesSaxon** [comment]
Hi @starkg,

What I have proposed still allows you to thread different electron
collections through the overlap removal or any subsequent algorithm.  I
simply don't follow here.

What is required now is to (a) change the systName, (b) link or unlink the
input collections in OR (in the example, "notice that the input algos for
jets and muons are empty"), and (c) change the subsequent algorithms to
look at the electronORAlgo or muonORAlgo or JetORAlgo.

What I have proposed indeed is that one would only change (a) the
m_systName, as you appear to intimate @mmilesi suggested.  But per Marco's
suggestion I would have to change (empty) the unvaried inputAlgos.  Finally
I think it is not intended that one need change the downstream algorithms.
JetORAlgo and MuonORAlgo should always contain the syst vectors for jets or
muons.  The existence of JetORAlgo or ElectronORAlgo vector shouldn't
depend on if I've run that syst or not.


Thanks a lot --

Jamie

**Comment by mmilesi** [comment]
Hi @JamesSaxon ,

I think you raised indeed some good points. I had a deeper look at the logic of the OR, and I think I have found a solution for both (b) and (c):

(b) : I added a simple function in `HelperFunctions`:

https://github.com/UCATLAS/xAODAnaHelpers/blob/FixNominalSysOverlapRemoval/xAODAnaHelpers/HelperFunctions.h#L360

that gets called in OR in case the `inputAlgo*` string is non empty: having in mind that the OR between the all-nominal containers will be always executed, the OR will be rerun only if the  `inputAlgo*` string is non empty AND there is at least one non-empty syst in the vector<string> of syst associated to the inputAlgo. Thus, having a non empty inputAlgo in the configuration will not cause any problem.

(c)
 I have realised that, for the OR, there is really no need to have a `m_outputAlgoElectrons` set.

In fact, **unlike the Selectors** (in the case the user sets a minimal/max requirement on the number of sel.objects), **the OR will never be set to skip entire events** (rather, it just resizes the object containers). 
This means that the input set of syst to the OR will not need to be resized in any case --> **input syst list = output syst list**.

For this reason, I changed the OR not to save in TStore any vector<string> of syst names, as it would be just redundant (and if we really wanted to have them in place, it would open up some difficulties in the implementation since we would like to have one `*ORAlgo` for every object, as you said, but at the same time make sure the vector<string> does not get stored to TStore more than once with the same name). 

To preserve backward compatibility, I kept the `m_outputAlgoElectrons, m_outputAlgoMuons,..` attributes in the class. However, they will be  in practice not used in the algorithm. The downstream algorithms should then pick up

`m_inputAlgoSystNames = "*Selector_Syst"`

or 

`m_inputAlgoSystNames = "*Calibrator_Syst"`  

, in case there are no selectors in the workflow.

Please see #275, which I recommend you to try out.

What do you think?

Marco

**Comment by JamesSaxon** [comment]
Hi @mmilesi --

Thanks so much for the careful thought and fast response on this.

I have one last concern, namely that I think that post overlap, one
*should* have
a single vector with *all* of the variations.

For instance, if some electron variation changes the selected jets (the
whole point of propagating these...), then I should be able to pick up see
that pick up that variation when I run e.g., a dijet maker.  If I have a
dijet maker after my OR, I want it to be able to listen to the variations
that came through the OR, without changing the post-OR "threading" when
varying an electron or a jet (only the enable systematics).

This is all the more relevant for the case where I run variations for
electrons and jets simultaneously (which I think should be able to work).
In that case, my dijet algorithm should look for a single list of
systematics to run over, from the OR tool.  In fact, as it stands, it's not
possible for me to tell it to pick up *both* containers -- not only because
I can't assign two vectors but also because each of them might include
"nominal" which would lead to a duplicate collection.

That leads to a very slight caveat in your (c), namely that it's not input
= output, but rather output = unique(inputElec + inputMuon + inputJets),
where I emphasize that it is unique, for the case of the nominal container
that exists in all of them.

So in short, I agree that having three copies of outputAlgo stored is
redundant, but that it would indeed be useful to have one.

How does this sound?


Best --

Jamie

**Comment by mmilesi** [comment]
Hi @JamesSaxon ,

I am not sure I got your point.

Why can't you just modify slightly your Dijet algo to have multiple `m_inputAlgo*` attributes (i.e., one for each object ), retrieve in it the 3:

```
m_inputAlgoSystNames_Electrons = "ElectronSelector_Syst"
m_inputAlgoSystNames_Jets = "JetSelector_Syst"
m_inputAlgoSystNames_Muons = "MuonSelector_Syst"
```

and, intenally to the Dijet code, merge them in a single vector<string>  with the entire objects' syst list - possibly with any redundancy eliminated?

Marco

**Comment by JamesSaxon** [comment]
Hi @mmilesi,

You're absolutely right that three inputs to all subsequent (event-level) algorithms would work.  I just think it's simpler to have the hand-off point (post-OR) from xAH to event-level analysis algos be a single set of systematics.  If you think the "three-vector pass" makes more sense, that's fine.

For completeness, I drew up one implementation of a unique (set) merge below -- it's easy in either spot.

Thanks again to for your consideration -- 

Jamie


```
  v.reserve(a.size() + b.size() + c.size());
  v.insert(v.end(), a.begin(), a.end());
  v.insert(v.end(), b.begin(), b.end());
  v.insert(v.end(), c.begin(), c.end());
  set<int> s(v.begin(), v.end());
  v.assign( s.begin(), s.end() );
```

**Comment by mmilesi** [comment]
Hi @JamesSaxon ,

can you please explain me why would you need to vary electrons and jets syst simultaneously? Sorry, this is just for my understanding (I cannot think of a typical use case off the top of my head :wink: ) and might help me to figure the scenario more clearly

**Comment by JamesSaxon** [comment]
Hi @mmilesi --

Two answers to this:

 1) I am not proposing simultaneous variation of jets and electrons -- just
the ability to run all systematics in one go (in parallel).
 2) It would also be nice to be able to run one or the other without
"relinking" multiple lines of algorithms.  I had thought of the algo names
as the "threads," and that one link should be enough to define the
"collection" of all the variations from the subsequent step.

Cheers --

Jamie

**Comment by mmilesi** [comment]
Hi @JamesSaxon ,

if I am not mistaken, the current version of #287 does not "break" the thread. Given a "general" workflow like the following:

```
1A. electron calibrator
     m_outputAlgoSystNames = "ElectronCalibrator_Syst"
1B. jet calibrator
     m_outputAlgoSystNames = "ElectronCalibrator_Syst"

2A. electron selector
     m_inputAlgoSystNames = "ElectronCalibrator_Syst"
     m_outputAlgoSystNames = "ElectronSelector_Syst"
2B. jet selector
     m_inputAlgoSystNames = "JetCalibrator_Syst"
     m_outputAlgoSystNames = "JetSelector_Syst"

3. Overlap removal
    m_inputAlgoElectrons = "ElectronSelector_Syst"
    m_inputAlgoJets = "JetSelector_Syst"
    (no need to have an output list for OR, see my comments above)
```

Then a following algorithm will be seeded by the xAOD containers outputted by the Overlap Removal module, and the systematic thread will follow instead from the selectors.
I want to stress that now this will work also in case the user is running the nominal case only!

 Basically, I do not think the OR should output any systematic list - rather, it will only need the input list(s) to correctly generate the combinatorial set of output containers. I'd leave it up to the user to merge in a unique list of syst in their own algorithm. I'd refrain from modifying too much the current structure, unless strictly needed.

Thanks a lot for all the feedback! I am happy that all the pieces start getting into a good shape :blush:

I'd appreciate if you could give an extensive try to #287 another time, just to make sure we haven't forgotten anything. 

Cheers

Marco

**Comment by mmilesi** [comment]
Closed by #275
